server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregationBuilder.java

```diff
@@ -347,16 +347,20 @@ public ZoneId timeZone() {
     protected final ValuesSourceAggregatorFactory doBuild(AggregationContext context, AggregatorFactory parent,
                                                           Builder subFactoriesBuilder) throws IOException {
         ValuesSourceConfig config = resolveConfig(context);
+
+        ValuesSourceAggregatorFactory factory;
         if (context.getValuesSourceRegistry().isRegistered(getRegistryKey())) {
             /*
             if the aggregation uses the values source registry, test if the resolved values source type is compatible with this aggregation.
             This call will throw if the mapping isn't registered, which is what we want.  Note that we need to throw from here because
             AbstractAggregationBuilder#build, which called this, will attempt to register the agg usage next, and if the usage is invalid
             that will fail with a weird error.
              */
-            context.getValuesSourceRegistry().getAggregator(getRegistryKey(), config);
+            Object aggregatorSupplier = context.getValuesSourceRegistry().getAggregator(getRegistryKey(), config);
```

Comment:
So, as you noted in your comments, having to use a bare `Object` here is a bit of a smell.  But, we don't need an interface to solve it, we already have a solution.  The way the generics are set up, we know the class we get out of the registry has to match the type on the registry key.  If you look at how the factories get the suppliers, they never have to cast it, because they know the key they're using, like this:
```
  DateHistogramAggregationSupplier aggregatorSupplier = context.getValuesSourceRegistry()	
            .getAggregator(DateHistogramAggregationBuilder.REGISTRY_KEY, config);
```
We can't do that here, because we don't know which type we're expecting.  But, we can do it from within `innerBuild`, since at that point we'll be in a specific aggregation.  So essentially we just want to pull the supplier fetching code up from the factory's `doCreateInternal` method to the builder's `internalBuild` method, and pass it on into the factory from there.  That still meets out validation need, it'll just throw from one level deeper if the mapping is bad, and should dodge all this bare `Object` business.